### PR TITLE
add migration utilities

### LINKED ISSUE #9396: checkpoint migration
## 🚀 Feature

Add upgrade functions to utilities for internal use. Checkpoints get upgraded automatically (when possible) when a user loads the checkpoint using `Trainer(resume_from_checkpoint)` or `Model.load_from_checkpoint`. 

### Motivation

Lightning changes over time with removals and additions, this includes checkpoint[ contents and structure](https://github.com/PyTorchLightning/pytorch-lightning/blob/3070a9ea6edb9648152218a79620f035d9bf719c/pytorch_lightning/core/saving.py#L167). When changes happen, we bake the upgrade logic into the code base at the appropriate place, but the danger is that the information why and when these changes were made [gets lost over time](https://github.com/PyTorchLightning/pytorch-lightning/blob/3070a9ea6edb9648152218a79620f035d9bf719c/pytorch_lightning/utilities/argparse_utils.py#L7). 

### Pitch

For each BC change we create an upgrade function that gets applied at the appropriate place.

```py
def upgrade_xyz_v1_2_0(checkpoint):
    # upgrades the checkpoint from a previous version to 1.2.0
    return checkpoint

def upgrade_abc_v1_3_8(checkpoint):
    # upgrades the checkpoint from previous version to 1.3.8
    return checkpoint


def upgrade(checkpoint)
    checkpoint = upgrade_xyz_v1_2_0(checkpoint)
    checkpoint = upgrade_abc_v1_3_8(checkpoint)
    return

# in Lightning:
ckpt = upgrade(pl_load(path))

```

Benefits with this approach are:
- each upgrade is documented individually
- central location for all upgrades, the order in which they are applied is fully transparent
- can unit test each upgrade individually! 

### Alternatives

keep as is

### Additional context

PRs that started this work: 
- #9166: legacy load context manager to patch Lightning for unpickling 
- #8558: upgrade functions

PRs that added checkpoint back-compatibility logic that can be avoided by this proposal:
- #11638
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7 @otaj

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds migration functions. When a checkpoint gets loaded, we migrate it automatically. 

Fixes #9396 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
Closing, there is too little value for this atm and the effort is quite large.

**Comment by carmocca** [comment]
Do you want to close the linked issue too? Or just dropping this implementation?

**Comment by awaelchli** [comment]
The linked issue can be kept open for the community if we want.
